libevmasm/Assembly.cpp

```diff
@@ -598,13 +599,13 @@ std::pair<std::shared_ptr<Assembly>, std::vector<std::string>> Assembly::fromJSO
 				{
 					solThrow(AssemblyImportException, "The key '" + dataItemID + "' inside '.data' is out of the supported integer range.");
 				}
+				solRequire(subId == index, AssemblyImportException, "Invalid order of keys in '.data'. Expect key '" + std::to_string(subId) + "' but got '" + dataItemID + "'.");
```

Comment:
Non-sequential order should be accepted. What we should reject are just gaps in numbering.

Reply:
Also, you don't really need `subId` here:
```suggestion
				solRequire(m_subs.size() == index, AssemblyImportException, "Invalid order of keys in '.data'. Expected key '" + std::to_string(m_subs.size()) + "' but got '" + dataItemID + "'.");
```

But really, it would be best to just read the stuff into a map, then check if indices are contiguous and then move the content to `m_subs`. That would remove the requirement for things to be in a specific order.

Reply:
I replaced by a map and added a test with random order, see https://github.com/ethereum/solidity/pull/13673/files#diff-3d83aa8de906c558ea951102bf94e11523ed10fc791104c052f87b942301118b